File: frontend/src/types/analytics.ts
export type AnalyticsTimeRange = '24h' | '7d' | '30d' | '90d' | 'all' | 'custom';

export type DataSufficiencyStatus = 'AVAILABLE' | 'INSUFFICIENT_DATA' | 'NO_RECORDS';

export interface MetricValue<T = number> {
  value: T | null;
  unit: string;
  status: DataSufficiencyStatus;
  sample_count: number;
  time_range: string;
  reason?: string | null;
}

export interface DurationDistribution {
  min_minutes: number | null;
  median_minutes: number | null;
  avg_minutes: number | null;
  max_minutes: number | null;
  sample_count: number;
  status: DataSufficiencyStatus;
  reason?: string | null;
}

export interface ResponseMilestoneTimeline {
  intake_to_acknowledgement: DurationDistribution;
  acknowledgement_to_situation: DurationDistribution;
  situation_to_plan_generation: DurationDistribution;
  plan_generation_to_officer_approval: DurationDistribution;
  approval_to_task_assignment: DurationDistribution;
  assignment_to_field_start: DurationDistribution;
  field_start_to_completion: DurationDistribution;
  incident_creation_to_resolution: DurationDistribution;
  total_samples: number;
}

export interface BottleneckInsight {
  bottleneck_type: string;
  severity: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  title: string;
  description: string;
  evidence: string;
  delay_impact_minutes?: number | null;
  affected_domain: string;
  affected_situation_id?: string | null;
  actionable_recommendation: string;
  detected_at: string;
}

export interface ResourceDemandCategory {
  category_name: string;
  total_stock: number;
  total_allocated: number;
  total_consumed: number;
  remaining_available: number;
  utilization_percentage: number;
  shortage_count: number;
  conflict_count: number;
}

export interface ResourceUtilizationAnalytics {
  time_range: string;
  total_resources_tracked: number;
  total_units_stock: number;
  total_units_allocated: number;
  total_units_consumed: number;
  overall_utilization_rate: number;
  categories: ResourceDemandCategory[];
  shortage_incidents_count: number;
  conflict_resolution_rate: number | null;
  status: DataSufficiencyStatus;
  reason?: string | null;
}

export interface ShelterAnalytics {
  total_shelters: number;
  total_capacity_beds: number;
  current_occupancy_beds: number;
  remaining_capacity_beds: number;
  occupancy_rate_percentage: number;
  full_shelters_count: number;
  unavailable_shelters_count: number;
  displaced_population_covered: number;
  unmet_shelter_demand_population: number;
  status: DataSufficiencyStatus;
}

export interface HealthcareAnalytics {
  total_facilities: number;
  total_beds_available: number;
  total_beds_occupied: number;
  bed_utilization_percentage: number;
  icu_beds_available: number;
  icu_beds_occupied: number;
  patient_evacuation_demand: number;
  critical_triage_demand: number;
  healthcare_shortages_count: number;
  status: DataSufficiencyStatus;
}

export interface VolunteerPerformanceAnalytics {
  registered_volunteers: number;
  active_responders: number;
  total_missions_assigned: number;
  missions_accepted: number;
  missions_in_progress: number;
  missions_completed: number;
  missions_blocked_or_failed: number;
  completion_rate_percentage: number;
  avg_mission_duration_minutes: number | null;
  skill_demand_breakdown: Record<string, number>;
  volunteer_shortages_count: number;
  status: DataSufficiencyStatus;
}

export interface FleetAnalytics {
  total_vehicles: number;
  active_fleet_missions: number;
  completed_fleet_missions: number;
  blocked_routes_reported: number;
  fleet_utilization_rate: number;
  avg_transit_minutes: number | null;
  vehicle_conflicts_count: number;
  status: DataSufficiencyStatus;
}

export interface ReplanningIntelligence {
  total_monitoring_events: number;
  impactful_events_detected: number;
  total_replans_executed: number;
  avg_replans_per_situation: number;
  affected_domains_distribution: Record<string, number>;
  replan_triggers_breakdown: Record<string, number>;
  avg_replan_resolution_minutes: number | null;
  status: DataSufficiencyStatus;
}

export interface IncidentComparisonMetric {
  group_name: string;
  incident_count: number;
  avg_response_minutes: number | null;
  avg_resolution_minutes: number | null;
  avg_replans: number;
  task_completion_rate: number;
  total_population_impacted: number;
}

export interface IncidentComparisonResponse {
  dimension: string;
  groups: IncidentComparisonMetric[];
  status: DataSufficiencyStatus;
  reason?: string | null;
}

export interface DecisionSupportSignal {
  signal_id: string;
  signal_type: string;
  severity: 'INFO' | 'WARNING' | 'CRITICAL';
  title: string;
  domain: string;
  evidence: string;
  affected_entity_id?: string | null;
  recommended_action: string;
  confidence: number;
  created_at: string;
}

export interface DecisionSupportResponse {
  signals: DecisionSupportSignal[];
  total_signals: number;
  critical_signals_count: number;
  warning_signals_count: number;
  generated_at: string;
}

export interface PostIncidentIntelligence {
  situation_id: string;
  title: string;
  emergency_type: string;
  severity_level: string;
  status: string;
  created_at: string;
  resolved_at?: string | null;
  closed_at?: string | null;
  total_duration_hours?: number | null;
  
  total_citizen_reports: number;
  initial_estimated_population: number;
  plan_versions_count: number;
  total_replans: number;
  participating_agents: string[];
  total_tasks_generated: number;
  tasks_completed: number;
  tasks_blocked_or_failed: number;
  task_completion_rate: number;
  
  resources_allocated_count: number;
  resources_consumed_count: number;
  shelters_activated_count: number;
  shelter_occupancy_peak: number;
  healthcare_referrals_count: number;
  volunteers_engaged_count: number;
  vehicles_deployed_count: number;
  route_disruptions_count: number;
  monitoring_events_count: number;
  
  intake_to_approval_minutes?: number | null;
  execution_duration_minutes?: number | null;
  identified_bottlenecks: string[];
  operational_improvements: string[];
  ai_summary_explanation?: string | null;
}

export interface EmergencyAnalyticsOverview {
  time_range: string;
  total_reports: number;
  total_situations: number;
  active_situations: number;
  resolved_situations: number;
  critical_incidents: number;
  high_severity_incidents: number;
  
  avg_acknowledgement_minutes: MetricValue<number>;
  avg_planning_minutes: MetricValue<number>;
  avg_approval_minutes: MetricValue<number>;
  avg_field_response_minutes: MetricValue<number>;
  avg_task_completion_minutes: MetricValue<number>;
  avg_incident_resolution_hours: MetricValue<number>;
  
  overall_task_completion_rate: number | null;
  overall_resource_utilization_rate: number;
  total_active_volunteers: number;
  total_active_fleet_missions: number;
  active_bottlenecks_count: number;
  active_decision_signals_count: number;
  generated_at: string;
}

File: backend/app/models/analytics.py
from enum import Enum
from typing import Optional, List, Dict, Any, Generic, TypeVar
from datetime import datetime
from pydantic import BaseModel, Field

T = TypeVar("T")


class AnalyticsTimeRange(str, Enum):
    LAST_24_HOURS = "24h"
    LAST_7_DAYS = "7d"
    LAST_30_DAYS = "30d"
    LAST_90_DAYS = "90d"
    ALL_TIME = "all"
    CUSTOM = "custom"


class DataSufficiencyStatus(str, Enum):
    AVAILABLE = "AVAILABLE"
    INSUFFICIENT_DATA = "INSUFFICIENT_DATA"
    NO_RECORDS = "NO_RECORDS"


class BottleneckType(str, Enum):
    ACKNOWLEDGEMENT_DELAY = "ACKNOWLEDGEMENT_DELAY"
    PLANNING_DELAY = "PLANNING_DELAY"
    OFFICER_APPROVAL_DELAY = "OFFICER_APPROVAL_DELAY"
    TASK_ASSIGNMENT_DELAY = "TASK_ASSIGNMENT_DELAY"
    FIELD_RESPONSE_DELAY = "FIELD_RESPONSE_DELAY"
    TASK_EXECUTION_DELAY = "TASK_EXECUTION_DELAY"
    RESOURCE_SHORTAGE = "RESOURCE_SHORTAGE"
    SHELTER_CAPACITY_SHORTAGE = "SHELTER_CAPACITY_SHORTAGE"
    HEALTHCARE_CAPACITY_SHORTAGE = "HEALTHCARE_CAPACITY_SHORTAGE"
    VOLUNTEER_SHORTAGE = "VOLUNTEER_SHORTAGE"
    TRANSPORT_SHORTAGE = "TRANSPORT_SHORTAGE"
    ROUTE_BLOCKAGE = "ROUTE_BLOCKAGE"
    REPEATED_REPLANNING = "REPEATED_REPLANNING"


class DecisionSignalType(str, Enum):
    RESOURCE_SHORTAGE_RISK = "RESOURCE_SHORTAGE_RISK"
    SHELTER_CAPACITY_RISK = "SHELTER_CAPACITY_RISK"
    HEALTHCARE_CAPACITY_RISK = "HEALTHCARE_CAPACITY_RISK"
    VOLUNTEER_AVAILABILITY_RISK = "VOLUNTEER_AVAILABILITY_RISK"
    TRANSPORT_CONGESTION_RISK = "TRANSPORT_CONGESTION_RISK"
    REPEATED_ROUTE_DISRUPTION = "REPEATED_ROUTE_DISRUPTION"
    APPROVAL_BOTTLENECK = "APPROVAL_BOTTLENECK"
    EXECUTION_DELAY_ALERT = "EXECUTION_DELAY_ALERT"
    REPLANNING_CASCADE_WARNING = "REPLANNING_CASCADE_WARNING"
    HIGH_UNRESOLVED_POPULATION_NEED = "HIGH_UNRESOLVED_POPULATION_NEED"


class MetricValue(BaseModel, Generic[T]):
    value: Optional[T] = None
    unit: str = ""
    status: DataSufficiencyStatus = DataSufficiencyStatus.AVAILABLE
    sample_count: int = 0
    time_range: str = "all"
    reason: Optional[str] = None


class DurationDistribution(BaseModel):
    min_minutes: Optional[float] = None
    median_minutes: Optional[float] = None
    avg_minutes: Optional[float] = None
    max_minutes: Optional[float] = None
    sample_count: int = 0
    status: DataSufficiencyStatus = DataSufficiencyStatus.AVAILABLE
    reason: Optional[str] = None


class ResponseMilestoneTimeline(BaseModel):
    intake_to_acknowledgement: DurationDistribution
    acknowledgement_to_situation: DurationDistribution
    situation_to_plan_generation: DurationDistribution
    plan_generation_to_officer_approval: DurationDistribution
    approval_to_task_assignment: DurationDistribution
    assignment_to_field_start: DurationDistribution
    field_start_to_completion: DurationDistribution
    incident_creation_to_resolution: DurationDistribution
    total_samples: int = 0


class BottleneckInsight(BaseModel):
    bottleneck_type: BottleneckType
    severity: str = "MEDIUM"  # LOW, MEDIUM, HIGH, CRITICAL
    title: str
    description: str
    evidence: str
    delay_impact_minutes: Optional[float] = None
    affected_domain: str
    affected_situation_id: Optional[str] = None
    actionable_recommendation: str
    detected_at: datetime


class ResourceDemandCategory(BaseModel):
    category_name: str
    total_stock: float
    total_allocated: float
    total_consumed: float
    remaining_available: float
    utilization_percentage: float
    shortage_count: int
    conflict_count: int


class ResourceUtilizationAnalytics(BaseModel):
    time_range: str
    total_resources_tracked: int
    total_units_stock: float
    total_units_allocated: float
    total_units_consumed: float
    overall_utilization_rate: float
    categories: List[ResourceDemandCategory] = Field(default_factory=list)
    shortage_incidents_count: int = 0
    conflict_resolution_rate: Optional[float] = None
    status: DataSufficiencyStatus = DataSufficiencyStatus.AVAILABLE
    reason: Optional[str] = None


class ShelterAnalytics(BaseModel):
    total_shelters: int = 0
    total_capacity_beds: float = 0.0
    current_occupancy_beds: float = 0.0
    remaining_capacity_beds: float = 0.0
    occupancy_rate_percentage: float = 0.0
    full_shelters_count: int = 0
    unavailable_shelters_count: int = 0
    displaced_population_covered: float = 0.0
    unmet_shelter_demand_population: float = 0.0
    status: DataSufficiencyStatus = DataSufficiencyStatus.AVAILABLE


class HealthcareAnalytics(BaseModel):
    total_facilities: int = 0
    total_beds_available: float = 0.0
    total_beds_occupied: float = 0.0
    bed_utilization_percentage: float = 0.0
    icu_beds_available: float = 0.0
    icu_beds_occupied: float = 0.0
    patient_evacuation_demand: int = 0
    critical_triage_demand: int = 0
    healthcare_shortages_count: int = 0
    status: DataSufficiencyStatus = DataSufficiencyStatus.AVAILABLE


class VolunteerPerformanceAnalytics(BaseModel):
    registered_volunteers: int = 0
    active_responders: int = 0
    total_missions_assigned: int = 0
    missions_accepted: int = 0
    missions_in_progress: int = 0
    missions_completed: int = 0
    missions_blocked_or_failed: int = 0
    completion_rate_percentage: float = 0.0
    avg_mission_duration_minutes: Optional[float] = None
    skill_demand_breakdown: Dict[str, int] = Field(default_factory=dict)
    volunteer_shortages_count: int = 0
    status: DataSufficiencyStatus = DataSufficiencyStatus.AVAILABLE


class FleetAnalytics(BaseModel):
    total_vehicles: int = 0
    active_fleet_missions: int = 0
    completed_fleet_missions: int = 0
    blocked_routes_reported: int = 0
    fleet_utilization_rate: float = 0.0
    avg_transit_minutes: Optional[float] = None
    vehicle_conflicts_count: int = 0
    status: DataSufficiencyStatus = DataSufficiencyStatus.AVAILABLE


class ReplanningIntelligence(BaseModel):
    total_monitoring_events: int = 0
    impactful_events_detected: int = 0
    total_replans_executed: int = 0
    avg_replans_per_situation: float = 0.0
    affected_domains_distribution: Dict[str, int] = Field(default_factory=dict)
    replan_triggers_breakdown: Dict[str, int] = Field(default_factory=dict)
    avg_replan_resolution_minutes: Optional[float] = None
    status: DataSufficiencyStatus = DataSufficiencyStatus.AVAILABLE


class IncidentComparisonMetric(BaseModel):
    group_name: str
    incident_count: int
    avg_response_minutes: Optional[float] = None
    avg_resolution_minutes: Optional[float] = None
    avg_replans: float = 0.0
    task_completion_rate: float = 0.0
    total_population_impacted: int = 0


class IncidentComparisonResponse(BaseModel):
    dimension: str  # emergency_type, severity_level, zone_or_district
    groups: List[IncidentComparisonMetric] = Field(default_factory=list)
    status: DataSufficiencyStatus = DataSufficiencyStatus.AVAILABLE
    reason: Optional[str] = None


class DecisionSupportSignal(BaseModel):
    signal_id: str
    signal_type: DecisionSignalType
    severity: str  # INFO, WARNING, CRITICAL
    title: str
    domain: str
    evidence: str
    affected_entity_id: Optional[str] = None
    recommended_action: str
    confidence: float = 1.0
    created_at: datetime


class DecisionSupportResponse(BaseModel):
    signals: List[DecisionSupportSignal] = Field(default_factory=list)
    total_signals: int = 0
    critical_signals_count: int = 0
    warning_signals_count: int = 0
    generated_at: datetime


class PostIncidentIntelligence(BaseModel):
    situation_id: str
    title: str
    emergency_type: str
    severity_level: str
    status: str
    created_at: datetime
    resolved_at: Optional[datetime] = None
    closed_at: Optional[datetime] = None
    total_duration_hours: Optional[float] = None
    
    # 12-factor debrief metrics
    total_citizen_reports: int = 0
    initial_estimated_population: int = 0
    plan_versions_count: int = 0
    total_replans: int = 0
    participating_agents: List[str] = Field(default_factory=list)
    total_tasks_generated: int = 0
    tasks_completed: int = 0
    tasks_blocked_or_failed: int = 0
    task_completion_rate: float = 0.0
    
    resources_allocated_count: int = 0
    resources_consumed_count: int = 0
    shelters_activated_count: int = 0
    shelter_occupancy_peak: float = 0.0
    healthcare_referrals_count: int = 0
    volunteers_engaged_count: int = 0
    vehicles_deployed_count: int = 0
    route_disruptions_count: int = 0
    monitoring_events_count: int = 0
    
    # Performance & delays
    intake_to_approval_minutes: Optional[float] = None
    execution_duration_minutes: Optional[float] = None
    identified_bottlenecks: List[str] = Field(default_factory=list)
    operational_improvements: List[str] = Field(default_factory=list)
    ai_summary_explanation: Optional[str] = None


class EmergencyAnalyticsOverview(BaseModel):
    time_range: str
    total_reports: int = 0
    total_situations: int = 0
    active_situations: int = 0
    resolved_situations: int = 0
    critical_incidents: int = 0
    high_severity_incidents: int = 0
    
    avg_acknowledgement_minutes: MetricValue[float]
    avg_planning_minutes: MetricValue[float]
    avg_approval_minutes: MetricValue[float]
    avg_field_response_minutes: MetricValue[float]
    avg_task_completion_minutes: MetricValue[float]
    avg_incident_resolution_hours: MetricValue[float]
    
    overall_task_completion_rate: Optional[float] = None
    overall_resource_utilization_rate: float = 0.0
    total_active_volunteers: int = 0
    total_active_fleet_missions: int = 0
    active_bottlenecks_count: int = 0
    active_decision_signals_count: int = 0
    generated_at: datetime
